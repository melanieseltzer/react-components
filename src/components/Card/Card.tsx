import React, { ReactNode } from 'react';
import styled from 'styled-components';

import Flex from '../Flex';

export interface Props {
  /** Must pass it a child. */
  children: ReactNode;
  /** Pass it any class to extend styling. */
  className?: string;
  /**
   * Pass it a valid box-shadow, otherwise default will be used.
   * Default: `0 5px 35px rgba(64, 67, 109, 0.25)`
   */
  boxShadow?: string;
  /**
   * Pass it a valid border-radius, otherwise default will be used.
   * Default: `10px`
   */
  borderRadius?: string;
}

const Card = (props: Props) => {
  const { children, className } = props;

  return (
    <CardWrapper className={className} {...props}>
      {children}
    </CardWrapper>
  );
};

const CardWrapper = styled(Flex.Item)`
  box-shadow: ${(props: Props) =>
    props.boxShadow ? props.boxShadow : `0 5px 35px rgba(64, 67, 109, 0.25)`};
  border-radius: ${(props: Props) =>
    props.borderRadius ? props.borderRadius : `10px`};
  padding: 1.25rem;
  margin: 1.25rem;
  transition: 0.2s all ease-out;
  transform: scale(1);
  &:hover {
    transform: scale(1.08);
  }
  @media (hover: none) {
    &:hover {
      transform: scale(1);
    }
  }
`;

export default Card;
