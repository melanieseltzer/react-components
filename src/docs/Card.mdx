---
name: Card
menu: Components
route: /card
---

import { Playground, Props } from 'docz';
import Card from '../components/Card/Card';
import Flex from '../components/Flex/Flex';

# Card

A super versatile Card component which is based on a Flex Item. Pass it any valid prop from Flex Item, in addition to a couple Card-specific props.

## Props

<Props of={Card} />

## Usage

### Default box-shadow

<Playground>
  <Flex>
    <Card flex="1 1 30%" background="tomato">
      Hi!
    </Card>
    <Card flex="1 1 30%" background="#fff">
      Hi!
    </Card>
    <Card flex="1 1 30%" background="lightsalmon">
      Hi!
    </Card>
  </Flex>
</Playground>

### Custom box-shadow

<Playground>
  <Card background="lightsalmon" boxShadow="0 2px 20px 0 lightsalmon">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce venenatis ut
    sem sit amet ullamcorper. Donec sollicitudin aliquam nisi id convallis.
    Proin quis felis lorem. Sed ac congue nisl, ut eleifend dui. Proin tempus
    pharetra imperdiet. Proin id justo nec lectus facilisis maximus quis eget
    mauris. Nunc et nibh orci.
  </Card>
</Playground>

### Custom border-radius

<Playground>
  <Card borderRadius="0">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce venenatis ut
    sem sit amet ullamcorper. Donec sollicitudin aliquam nisi id convallis.
    Proin quis felis lorem. Sed ac congue nisl, ut eleifend dui. Proin tempus
    pharetra imperdiet. Proin id justo nec lectus facilisis maximus quis eget
    mauris. Nunc et nibh orci.
  </Card>
</Playground>

### Flexible interior content

<Playground>
  <Card>
    <h2>Subtitle</h2>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce venenatis
      ut sem sit amet ullamcorper. Donec sollicitudin aliquam nisi id convallis.
      Proin quis felis lorem. Sed ac congue nisl, ut eleifend dui. Proin tempus
      pharetra imperdiet. Proin id justo nec lectus facilisis maximus quis eget
      mauris. Nunc et nibh orci.
    </p>
    <a href="https://github.com">GitHub</a>
  </Card>
</Playground>
