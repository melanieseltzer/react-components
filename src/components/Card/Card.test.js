import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Card from '.';

afterEach(cleanup);

describe('Card component', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <Card>
        <div>Hi</div>
      </Card>
    );
    expect(container).toMatchSnapshot();
  });

  it('should be rendered with at least one child component', () => {
    const { container } = render(
      <Card>
        <div>Hi</div>
      </Card>
    );
    expect(container.firstChild.hasChildNodes).toBeTruthy();
  });

  it('should render with a class of test', () => {
    const { container } = render(
      <Card className="test">
        <div>Hi</div>
      </Card>
    );
    expect(container.firstChild.classList.contains('test')).toBe(true);
  });

  it('should render with default box shadow if not passed explicitely', () => {
    const { container } = render(
      <Card>
        <div>Hi</div>
      </Card>
    );
    expect(container.firstChild).toHaveStyleRule(
      'box-shadow',
      '0 5px 35px rgba(64,67,109,0.25)'
    );
  });

  it('should render with default border radius if not passed explicitely', () => {
    const { container } = render(
      <Card>
        <div>Hi</div>
      </Card>
    );
    expect(container.firstChild).toHaveStyleRule('border-radius', '10px');
  });
});
