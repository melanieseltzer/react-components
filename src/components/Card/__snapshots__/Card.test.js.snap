// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches the snapshot 1`] = `
.c0 {
  box-shadow: 0 5px 35px rgba(64,67,109,0.25);
  border-radius: 10px;
  padding: 1.25rem;
  margin: 1.25rem;
  -webkit-transition: 0.2s all ease-out;
  transition: 0.2s all ease-out;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c0:hover {
  -webkit-transform: scale(1.02);
  -ms-transform: scale(1.02);
  transform: scale(1.02);
}

@media (hover:none) {
  .c0:hover {
    -webkit-transform: scale(1);
    -ms-transform: scale(1);
    transform: scale(1);
  }
}

<div>
  <div
    class="c0 "
  >
    <div>
      Hi
    </div>
  </div>
</div>
`;
